plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -251,5 +261,49 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
     override fun accessInvariants(v: Exp): List<Exp> =
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter(v) }
 
+    // Note: this function will replace accessInvariants when nested unfold will be implemented
+    override fun predicateAccessInvariants(v: Exp): List<Exp> = listOf(Exp.PredicateAccess(name, listOf(v)))
+
     override fun provenInvariants(v: Exp) = listOf(subtypeInvariant(v))
+
+    private val thisEmbedding = VariableEmbedding(ThisReceiverName, this)
+    val classPredicate: Predicate
```

Comment:
This is doing a lot of work in a getter that may look cheap to call.  Would initialising this once make sense?  (Perhaps have a function that does all this and then just assign it to the property on init?)

Reply:
I am not sure that this is possible, the predicate of a class depends on the fields which are not necessarily available during the initialization

Reply:
Could you write a comment somewhere about the expected order?  I see what you mean, I forgot we initialise fields stage-by-stage.  Perhaps there's a specific function that should be called at the right moment?

Reply:
Actually, we already have a `setFields` function.  Maybe `predicate` should be initialised from there?